### PR TITLE
control_plane: revise compute_hook locking (don't serialise all calls)

### PR DESCRIPTION
**Pull request by jcsp**
## Problem

- Previously, an async mutex was held for the duration of `ComputeHook::notify`.  This served multiple purposes:
  - Ensure updates to a given tenant are sent in the proper order
  - Prevent concurrent calls into neon_local endpoint updates in test environments (neon_local is not safe to call concurrently)
  - Protect the inner ComputeHook::state hashmap that is used to calculate when to send notifications.

This worked, but had the major downside that while we're waiting for a compute hook request to the control plane to succeed, we can't notify about any other tenants.  Notifications block progress of live migrations, so this is a problem.

## Summary of changes

- Protect `ComputeHook::state` with a sync  lock instead of an async lock
- Use a separate async lock ( `ComputeHook::neon_local_lock` ) for preventing concurrent calls into neon_local, and only take this in the neon_local code path.
- Add per-tenant async locks in ShardedComputeHookTenant, and use these to ensure that only one remote notification can be sent at once per tenant.  If several shards update concurrently, their updates will be coalesced.
- Add an explicit semaphore that limits concurrency of calls into the cloud control plane.

## Checklist before requesting a review

- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics? if so did you add the relevant metrics to the dashboard?
- [ ] If this PR requires public announcement, mark it with /release-notes label and add several sentences in this section.

## Checklist before merging

- [ ] Do not forget to reformat commit message to not include the above checklist

### CODE REVIEW

File: control_plane/attachment_service/src/compute_hook.rs [hunk 1]
```diff
@@ -340,30 +386,38 @@ impl ComputeHook {
...
             // until it does.
             tracing::info!("Tenant isn't yet ready to emit a notification");
             return Ok(());
         };
 
+        // Finish acquiring the tenant's async lock: this future was created inside the self.state
+        // lock above, so we are guaranteed to get this lock in the same order as callers took
+        // that lock.  This ordering is essential: the cloud control plane must end up with the
+        // same end state for the tenant that we see.
+        let _guard = lock_fut.await;
```
Comment (problame) [inline control_plane/attachment_service/src/compute_hook.rs:419]: "this lock" and "that lock"; are they referring to two different locks or the same lock?

Also, if it's about two different locks, I don't think this is correct: 
- task A gets preempted after producing the future
- task B gets preempted after producing the future
- task B does this here `lock_fut.await`
- task A does this here `lock_fut.await`

**Comment by jcsp** [comment]
This didn't reduce down as neatly as I had hoped, but I  think the logic is sound: we try to acquire a ComputeHookTenant's async lock before sending an  update, and if we can't acquire it we drop out, wait for the lock, and try again.   The  request  we send  is always derived from latest state of the ComputeHookTenant, so there is no longer any  sensitivity to which order notify() calls complete in.

I considered using an async lock for ComputeHook::state to make the  control flow a bit less tortured, but on balance I prefer to  use a sync lock so that  the compiler enforces that we do  not hold it while waiting for IO.

File: control_plane/attachment_service/src/compute_hook.rs [hunk 2]
```diff
@@ -343,42 +433,65 @@ impl ComputeHook {
...
+                return Ok(());
             }
+            MaybeSendResult::AwaitLock(lock) => {
+                let send_locked = lock.lock_owned().await;
+
+                let locked = self.state.lock().unwrap();
+                let Some(tenant) = locked.get(&tenant_shard_id.tenant_id) else {
+                    return Ok(());
+                };
+                match tenant.maybe_send(tenant_shard_id.tenant_id, Some(send_locked)) {
```
Comment (problame) [inline control_plane/attachment_service/src/compute_hook.rs:468]: The lock order is different than above.
Above, we lock self.state first, then call tenant.maybe_send(), and maybe_send() internally locks the same lock that we get here in variable `lock`.

It's not a deadlock because `maybe_send()` does a `try_` lock operation.

Smells bad, but I think it won't deadlock.

I think the coalescing should be done using heavier_once_cell, but, that can be cleaned up later.

File: control_plane/attachment_service/src/compute_hook.rs [hunk 3]
```diff
@@ -37,9 +41,17 @@ impl ComputeHookTenant {
...
                 shard_count: tenant_shard_id.shard_count,
+                send_lock: Arc::default(),
             })
         } else {
-            Self::Unsharded(node_id)
+            Self::Unsharded((node_id, Arc::default()))
+        }
+    }
+
+    fn get_lock(&self) -> &Arc<tokio::sync::Mutex<ComputeHookNotifyRequest>> {
```
Comment (problame) [inline control_plane/attachment_service/src/compute_hook.rs:51]: Method should be called `get_send_lock()` for clarity at callers.

File: control_plane/attachment_service/src/compute_hook.rs [hunk 4]
```diff
@@ -343,42 +433,65 @@ impl ComputeHook {
...
-                let tenant = e.into_mut();
-                tenant.update(tenant_shard_id, stripe_size, node_id);
-                tenant
+        // Process result: we may get an update to send, or we may have to wait for a lock
+        // before trying again.
+        let (request, mut guard) = match maybe_send_result {
+            MaybeSendResult::Noop => {
+                return Ok(());
             }
+            MaybeSendResult::AwaitLock(lock) => {
```
Comment (problame) [inline control_plane/attachment_service/src/compute_hook.rs:461]: Please pick more expressive names for `guard` and `lock`.
E.g.,

* `lock` => `send_lock`
* `guard` => `send_lock_guard`

**Review by problame — APPROVED** [review]
I think this code has no locking bugs, but, it's all quite brittle / hard to follow.

My only strong request for chagnes is the https://github.com/neondatabase/neon/pull/7088#discussion_r1553933078

The naming changes would be pretty nice, but, if you don't have time for them, I can do them in a follow-up.

File: control_plane/attachment_service/src/compute_hook.rs (hunk 2 above)
Comment (jcsp) [reply control_plane/attachment_service/src/compute_hook.rs]: Right, it's a bit brittle.  I've added a comment to explain/justify the lock order at both places we take the send lock in cf1b476c1258d2428228aca848255fdf2cfb5007.

**Comment by jcsp** [comment]
Comments on names/structs are all quite reasonable, amended in cf1b476c1258d2428228aca848255fdf2cfb5007
